R/filter.R

```diff
@@ -0,0 +1,104 @@
+#' Filter data by levels of numeric variables
+#'
+#' Filter a dataframe by some boolean condition of one numeric variable
+#' and returns the result along with tidyverse code used to generate it.
+#'
+#' @param data a dataframe or survey design object to filter
+#' @param var character of the column in `data` to filter by
+#' @param op  a logical operator of "<=", "<", ">=", ">", "==" or "!="
+#'        for the boolean condition
+#' @param num a number for which the \code{op} applies to
+#' @return filtered dataframe with tidyverse code attached
+#' @seealso \code{\link{code}}
+#' @rdname filter_num
+#' @examples
+#' filtered <- filter_num(iris, var = "Sepal.Length", op = "<=", num = 5)
+#' cat(code(filtered))
+#' head(filtered)
+#'
+#' library(survey)
+#' data(api)
+#' svy <- svydesign(~ dnum + snum, weights = ~pw, fpc = ~ fpc1 + fpc2, data = apiclus2)
+#' svy_filtered <- filter_num(svy, var = "api00", op = "<", num = 700)
+#' cat(code(svy_filtered))
+#'
+#' @author Owen Jin, Tom Elliott, Stephen Su
+#' @export
+#' @md
+filter_num <- function(data, var,
+                       op = c("<=", "<", ">=", ">", "==", "!="),
+                       num) {
+    op <- rlang::arg_match(op)
+    expr <- rlang::enexpr(data)
+    if (is_survey(data) && !inherits(data, "tbl_svy")) {
+        expr <- rlang::expr(!!expr %>% srvyr::as_survey())
+    }
+    filter_lib <- rlang::sym(ifelse(is_survey(data), "srvyr", "dplyr"))
+    expr <- rlang::expr(!!expr %>%
+        `::`(!!filter_lib, filter)((!!op)(!!rlang::sym(var), !!num)))
```

Comment:
This is a very clever solution, but it's too complicated and difficult to figure out what's going on if you don't already know what the code is doing. The code is cleaner and more concise using an `if (survey) ...expression... else ...expression...` as I used in the [example the other day](https://github.com/iNZightVIT/iNZightTools/commit/c815d1050d9d872314fe958023c5775a344f6be0#r92972239). 

Trying to debug this kind of code in the future for me (or, more likely, someone else) will be very confusing - especially once you get into the more complicated methods. It's almost always better to write simple rather than clever code unless you have a good reason to do so, and this is especially true in collaborative environments like this one.

Reply:
For this part: `(!!op)(!!rlang::sym(var), !!num)` it would be better - if possible - to assign this to a variable with a comment explaining what it's doing, e.g., `# create filtering expression`, and the pass that into the function instead of having a bracket-festival on a single line.

Reply:
I haven't test run the following code yet, but do you think it makes the code more readable?

```r
filter_num <- function(data, var,
                       op = c("<=", "<", ">=", ">", "==", "!="),
                       num) {
    op <- rlang::arg_match(op)
    expr <- rlang::enexpr(data)
    if (is_survey(data) && !inherits(data, "tbl_svy")) {
        expr <- rlang::expr(!!expr %>% srvyr::as_survey())
    }
    fn <- rlang::expr(`::`(!!rlang::sym(ifelse(is_survey(data), "srvyr", "dplyr")), filter)) ## Make package_name::filter
    filtering <- rlang::expr((!!op)(!!rlang::sym(var), !!num)) ## Turn into "var op num" expression form
    expr <- rlang::expr(!!expr %>% !!fn(!!filtering))
    eval_code(expr)
}
```

Reply:
This is better but probably best saved for more complex situations where it *does* make sense to create function names dynamically (e.g., if this replaces a whole bunch of if-else-if-else statements, for example). In this case, it is still far cleaner to just use
```r
filter_expression <- rlang::expr((!!op)(!!rlang::sym(var), !!num))
if (is_survey(data) && !inherits(data, "tbl_svy")) {
    expr <- rlang::expr(!!expr %>% srvyr::as_survey()) %>%
        srvyr::filter(!!filter_expression))
} else {
    expr <- rlang::expr(!!expr %>% dplyr::filter(!!filter_expression))
}
```